runtime/cudaq/cudaq.cpp

```diff
@@ -15,80 +15,150 @@
 #include "cudaq/platform.h"
 #include "cudaq/utils/registry.h"
+#include "distributed/mpi_plugin.h"
 #include <dlfcn.h>
+#include <filesystem>
 #include <map>
 #include <regex>
 #include <signal.h>
 #include <string>
 #include <vector>
-
-#ifdef CUDAQ_HAS_MPI
-#include <mpi.h>
-
 namespace nvqir {
 void tearDownBeforeMPIFinalize();
 void setRandomSeed(std::size_t);
 } // namespace nvqir
 
 namespace cudaq::mpi {
+cudaq::MPIPlugin *getMpiPlugin(bool unsafe) {
+  // Locate and load the MPI comm plugin.
+  // Rationale: we don't want to explicitly link `libcudaq.so` against any
+  // specific MPI implementation for compatibility. Rather, MPI functionalities
+  // are encapsulated inside a runtime-loadable plugin.
+  static std::unique_ptr<cudaq::MPIPlugin> g_plugin;
+  if (!g_plugin) {
+    // Search priority:
+    // (1) Environment variable take precedence
+    // (2) Built-in comm plugin (e.g., docker container or build from source
+    // with MPI)
+    // (3) mpi4py-based wrapper
+    const char *mpiLibPath = std::getenv("CUDAQ_MPI_COMM_LIB");
+    if (mpiLibPath) {
+      // The user has set the environment variable.
+      cudaq::info("Load MPI comm plugin from CUDAQ_MPI_COMM_LIB environment "
+                  "variable at '{}'",
+                  mpiLibPath);
+      g_plugin = std::make_unique<cudaq::MPIPlugin>(mpiLibPath);
+    } else {
+      // Try locate MPI plugins in the install directory
+      std::filesystem::path cudaqLibPath{cudaq::getCUDAQLibraryPath()};
+      const auto pluginsPath = cudaqLibPath.parent_path() / "plugins";
+#if defined(__APPLE__) && defined(__MACH__)
+      const std::string libSuffix = "dylib";
+#else
+      const std::string libSuffix = "so";
+#endif
+      // The builtin (native) plugin if present
+      const auto pluginLibFile =
+          pluginsPath / fmt::format("libcudaq-comm-plugin.{}", libSuffix);
+      // The mpi4py-based plugin
+      const auto pyPluginLibFile =
+          pluginsPath / fmt::format("libcudaq-py-comm-plugin.{}", libSuffix);
+      if (std::filesystem::exists(pluginLibFile)) {
+        cudaq::info("Load builtin MPI comm plugin from  at '{}'",
+                    pluginLibFile.c_str());
+        g_plugin = std::make_unique<cudaq::MPIPlugin>(pluginLibFile.c_str());
+      } else if (std::filesystem::exists(pyPluginLibFile)) {
+        cudaq::info("Try loading mpi4py MPI comm plugin from  at '{}'",
+                    pyPluginLibFile.c_str());
+        g_plugin = std::make_unique<cudaq::MPIPlugin>(pyPluginLibFile.c_str());
+        // With mpi4py plugin, we need to check if it is actually working.
+        // If mpi4py is not present at runtime, we cannot use this plugin.
+        if (!g_plugin->isValid()) {
+          cudaq::info("Failed to load mpi4py MPI comm plugin (mpi4py is not "
+                      "available).");
+          // Don't use it since mpi4py is not available.
+          g_plugin.reset();
+        }
+      }
+    }
+  }
+  if (!g_plugin) {
+    if (unsafe)
```

Comment:
so if I compile and then run an CUDAQ MPI-enabled code without `mpiexec`, then I'll get an exception thrown always. But I think the MPI standard allows for compiling an MPI code and running it apart from `mpiexec` and the result should be a code assuming a single rank. https://www.mcs.anl.gov/research/projects/mpi/mpi-standard/mpi-report-2.0/node112.htm (maybe I'm wrong about my interpretation here).